semantic_model_generator/data_processing/cte_utils.py

```diff
@@ -77,43 +80,107 @@ def remove_ltable_cte(sql_w_ltable_cte: str) -> str:
     return sql_without_logical_cte  # type: ignore [no-any-return]
 
 
-def _generate_cte_for(table: semantic_model_pb2.Table) -> str:
+def _validate_col(column: semantic_model_pb2.Column) -> None:
+    if " " in column.name.strip():
+        raise ValueError(
+            f"Please do not include spaces in your column name: {column.name}"
+        )
+    if column.data_type.upper() in OBJECT_DATATYPES:
+        raise ValueError(
+            f"We do not support object datatypes in the semantic model. Col {column.name} has data type {column.data_type}. Please remove this column from your semantic model or flatten it to non-object type."
+        )
+
+
+def validate_all_cols(table: semantic_model_pb2.Table) -> None:
+    for column in table.columns:
+        _validate_col(column)
+
+
+def _get_col_expr(column: semantic_model_pb2.Column) -> str:
+    """Return column expr in SQL format.
+    Raise errors if columns is of OBJECT_DATATYPES, which we do not support today."""
+    return (
+        f"{column.expr.strip()} as {column.name.strip()}"
+        if column.expr.strip().lower() != column.name.strip().lower()
+        else f"{column.expr.strip()}"
+    )
+
+
+def _generate_cte_for(
+    table: semantic_model_pb2.Table, columns: List[semantic_model_pb2.Column]
+) -> str:
     """
     Returns a CTE representing a logical table that selects 'col' columns from 'table'.
     """
 
-    def _get_col_expr(column: semantic_model_pb2.Column) -> str:
-        return (
-            f"{column.expr} as {column.name}"
-            if column.expr.lower() != column.name.lower()
-            else f"{column.expr}"
+    if len(columns) == 0:
+        raise ValueError("Please include at least one column to generate CTE on.")
+    else:
+        expr_columns = [_get_col_expr(col) for col in columns]
+        cte = f"WITH {logical_table_name(table)} AS (\n"
+        cte += "SELECT \n"
+        cte += ",\n".join(expr_columns) + "\n"
+        cte += f"FROM {fully_qualified_table_name(table.base_table)}"
+        cte += ")"
+        return cte
+
+
+def _generate_non_agg_cte(table: semantic_model_pb2.Table) -> str:
+    """
+    Returns a CTE representing a logical table that selects 'col' columns from 'table' except for aggregation columns.
+    """
+    filtered_cols = [col for col in table.columns if not is_aggregation_expr(col)]
+    return _generate_cte_for(table, filtered_cols)
```

Comment:
we're guaranteed to always have non-agg columns because we'll add the underlying columns in the agg-col expression to the semantic model, right?

Can you add a comment to this effect here?

Reply:
would be great if we can fail gracefully if there are no columns at all.

Reply:
updated